js/Update.elm

```diff
@@ -543,202 +541,153 @@ changeStage stage model =
 -- HELPER UPDATERS
 
 
-updateIfWelcome :
-    (WelcomeModel -> ( AppModel, Cmd Msg ))
-    -> AppModel
-    -> ( AppModel, Cmd Msg )
-updateIfWelcome upd model =
-    case model of
-        WelcomeScreen m ->
-            upd m
-
-        _ ->
-            (Debug.log
-                ("Tried running update function "
-                    ++ toString upd
-                    ++ " during "
-                    ++ toString model
-                )
-            )
-                ( model, Cmd.none )
+type alias Lens a b s t =
+    { get : s -> Maybe a
+    , set : b -> s -> t
+    }
 
 
-tryUpdateWelcome :
-    (WelcomeModel -> ( WelcomeModel, Cmd Msg ))
-    -> AppModel
-    -> ( AppModel, Cmd Msg )
-tryUpdateWelcome upd =
-    updateIfWelcome <|
-        \m ->
-            let
-                ( newM, cmd ) =
-                    upd m
-            in
-                ( WelcomeScreen newM, cmd )
-
-
-updateIfGame :
-    (GameModel -> ( AppModel, Cmd Msg ))
-    -> AppModel
-    -> ( AppModel, Cmd Msg )
-updateIfGame upd model =
-    case model of
-        Game m ->
-            upd m
-
-        _ ->
-            (Debug.log
-                ("Tried running update function "
-                    ++ toString upd
-                    ++ " during "
-                    ++ toString model
-                )
-            )
-                ( model, Cmd.none )
+type alias EffLens a s =
+    Lens a (Eff a) s (Eff s)
 
 
-tryUpdateGame :
-    (GameModel -> ( GameModel, Cmd Msg ))
-    -> AppModel
-    -> ( AppModel, Cmd Msg )
-tryUpdateGame upd =
-    updateIfGame <|
-        \m ->
-            let
-                ( newM, cmd ) =
-                    upd m
-            in
-                ( Game newM, cmd )
-
-
-updateIfReady :
-    (ReadyModel -> GameModel -> ( GameModel, Cmd Msg ))
-    -> GameModel
-    -> ( GameModel, Cmd Msg )
-updateIfReady upd model =
-    case model.stage of
-        ReadyStage m ->
-            upd m model
-
-        _ ->
-            (Debug.log
-                ("Tried running update function "
-                    ++ toString upd
-                    ++ " during "
-                    ++ toString model.stage
-                )
-                identity
-            )
-                ( model, Cmd.none )
-
-
-tryUpdateReady :
-    (ReadyModel -> ( ReadyModel, Cmd Msg ))
-    -> GameModel
-    -> ( GameModel, Cmd Msg )
-tryUpdateReady upd =
-    updateIfReady <|
-        \m model ->
-            let
-                ( newM, cmd ) =
-                    upd m
-            in
-                ( { model | stage = ReadyStage newM }, cmd )
+goIn : EffLens a b -> EffLens b c -> EffLens a c
```

Comment:
Could you put some comments on some of these confusing types? I have a pretty hard time understanding what this type signature means.